# Predict survived Titanic dataset by using Logistic Regression
### We use logistic regression to predict binary outcome (0/1).
### In this case we use logistic regression to predict survived of Titanic dataset. Titanic dataset is dataset in library of R. 

In [1]:
install.packages("titanic")

Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done



## Load Titanic Library

In [2]:
library(titanic)

## Preview dataset
### When we preview dataset, we found that dataset have missing value in some column such as "Age" column.

In [3]:
View(titanic_train)

A data.frame: 891 × 12 
 
 PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked 
 <int> <int> <int> <chr> <chr> <dbl> <int> <int> <chr> <dbl> <chr> <chr> 
 
 
 1 0 3 Braund, Mr. Owen Harris male 22 1 0 A/5 21171 7.2500 S 
 2 1 1 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38 1 0 PC 17599 71.2833 C85 C 
 3 1 3 Heikkinen, Miss. Laina female 26 0 0 STON/O2. 3101282 7.9250 S 
 4 1 1 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35 1 0 113803 53.1000 C123 S 
 5 0 3 Allen, Mr. William Henry male 35 0 0 373450 8.0500 S 
 6 0 3 Moran, Mr. James male NA 0 0 330877 8.4583 Q 
 7 0 1 McCarthy, Mr. Timothy J male 54 0 0 17463 51.8625 E46 S 
 8 0 3 Palsson, Master. Gosta Leonard male 2 3 1 349909 21.0750 S 
 9 1 3 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27 0 2 347742 11.1333 S 
 10 1 2 Nasser, Mrs. Nicholas (Adele Achem) female 14 1 0 237736 30.0708 C 
 11 1 3 Sandstrom, Miss. Marguerite Rut female 4 1 1 PP 9549 16.7000 G6 S 
 12 1 1 Bonnell, Miss. Elizabeth female 58 0 0 113783 26.5500 C103 S 
 13 0 3 Saundercock, Mr. William Henry male 20 0 0 A/5. 2151 8.0500 S 
 14 0 3 Andersson, Mr. Anders Johan male 39 1 5 347082 31.2750 S 
 15 0 3 Vestrom, Miss. Hulda Amanda Adolfina female 14 0 0 350406 7.8542 S 
 16 1 2 Hewlett, Mrs. (Mary D Kingcome) female 55 0 0 248706 16.0000 S 
 17 0 3 Rice, Master. Eugene male 2 4 1 382652 29.1250 Q 
 18 1 2 Williams, Mr. Charles Eugene male NA 0 0 244373 13.0000 S 
 19 0 3 Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele) female 31 1 0 345763 18.0000 S 
 20 1 3 Masselmani, Mrs. Fatima female NA 0 0 2649 7.2250 C 
 21 0 2 Fynney, Mr. Joseph J male 35 0 0 239865 26.0000 S 
 22 1 2 Beesley, Mr. Lawrence male 34 0 0 248698 13.0000 D56 S 
 23 1 3 McGowan, Miss. Anna "Annie" female 15 0 0 330923 8.0292 Q 
 24 1 1 Sloper, Mr. William Thompson male 28 0 0 113788 35.5000 A6 S 
 25 0 3 Palsson, Miss. Torborg Danira female 8 3 1 349909 21.0750 S 
 26 1 3 Asplund, Mrs. Carl Oscar (Selma Augusta Emilia Johansson) female 38 1 5 347077 31.3875 S 
 27 0 3 Emir, Mr. Farred Chehab male NA 0 0 2631 7.2250 C 
 28 0 1 Fortune, Mr. Charles Alexander male 19 3 2 19950 263.0000 C23 C25 C27 S 
 29 1 3 O'Dwyer, Miss. Ellen "Nellie" female NA 0 0 330959 7.8792 Q 
 30 0 3 Todoroff, Mr. Lalio male NA 0 0 349216 7.8958 S 
 ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ 
 862 0 2 Giles, Mr. Frederick Edward male 21 1 0 28134 11.5000 S 
 863 1 1 Swift, Mrs. Frederick Joel (Margaret Welles Barron) female 48 0 0 17466 25.9292 D17 S 
 864 0 3 Sage, Miss. Dorothy Edith "Dolly" female NA 8 2 CA. 2343 69.5500 S 
 865 0 2 Gill, Mr. John William male 24 0 0 233866 13.0000 S 
 866 1 2 Bystrom, Mrs. (Karolina) female 42 0 0 236852 13.0000 S 
 867 1 2 Duran y More, Miss. Asuncion female 27 1 0 SC/PARIS 2149 13.8583 C 
 868 0 1 Roebling, Mr. Washington Augustus II male 31 0 0 PC 17590 50.4958 A24 S 
 869 0 3 van Melkebeke, Mr. Philemon male NA 0 0 345777 9.5000 S 
 870 1 3 Johnson, Master. Harold Theodor male 4 1 1 347742 11.1333 S 
 871 0 3 Balkic, Mr. Cerin male 26 0 0 349248 7.8958 S 
 872 1 1 Beckwith, Mrs. Richard Leonard (Sallie Monypeny) female 47 1 1 11751 52.5542 D35 S 
 873 0 1 Carlsson, Mr. Frans Olof male 33 0 0 695 5.0000 B51 B53 B55 S 
 874 0 3 Vander Cruyssen, Mr. Victor male 47 0 0 345765 9.0000 S 
 875 1 2 Abelson, Mrs. Samuel (Hannah Wizosky) female 28 1 0 P/PP 3381 24.0000 C 
 876 1 3 Najib, Miss. Adele Kiamie "Jane" female 15 0 0 2667 7.2250 C 
 877 0 3 Gustafsson, Mr. Alfred Ossian male 20 0 0 7534 9.8458 S 
 878 0 3 Petroff, Mr. Nedelio male 19 0 0 349212 7.8958 S 
 879 0 3 Laleff, Mr. Kristo male NA 0 0 349217 7.8958 S 
 880 1 1 Potter, Mrs. Thomas Jr (Lily Alexenia Wilson) female 56 0 1 11767 83.1583 C50 C 
 881 1 2 Shelley, Mrs. William (Imanita Parrish Hall) female 25 0 1 230433 26.0000 S 
 882 0 3 Markun, Mr. Johann male 33 0 0 349257 7.8958 S 
 883 0 3 Dahlberg, Miss. Gerda Ulrika female 22 0 0 7552 10.5167 S 
 884 0 2 Banfield, Mr. Frederick James male 28 0 0 C.A./SOTON 34068 

## Number of rows dataset

In [4]:
nrow(titanic_train)

891

## Drop missing value from dataset
### We should drop missing value before put dataset to model.

In [5]:
titanic_train <- na.omit(titanic_train)
titanic_train

A data.frame: 714 × 12 
 
 PassengerId Survived Pclass Name Sex Age SibSp Parch Ticket Fare Cabin Embarked 
 <int> <int> <int> <chr> <chr> <dbl> <int> <int> <chr> <dbl> <chr> <chr> 
 
 
 1 1 0 3 Braund, Mr. Owen Harris male 22 1 0 A/5 21171 7.2500 S 
 2 2 1 1 Cumings, Mrs. John Bradley (Florence Briggs Thayer) female 38 1 0 PC 17599 71.2833 C85 C 
 3 3 1 3 Heikkinen, Miss. Laina female 26 0 0 STON/O2. 3101282 7.9250 S 
 4 4 1 1 Futrelle, Mrs. Jacques Heath (Lily May Peel) female 35 1 0 113803 53.1000 C123 S 
 5 5 0 3 Allen, Mr. William Henry male 35 0 0 373450 8.0500 S 
 7 7 0 1 McCarthy, Mr. Timothy J male 54 0 0 17463 51.8625 E46 S 
 8 8 0 3 Palsson, Master. Gosta Leonard male 2 3 1 349909 21.0750 S 
 9 9 1 3 Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg) female 27 0 2 347742 11.1333 S 
 10 10 1 2 Nasser, Mrs. Nicholas (Adele Achem) female 14 1 0 237736 30.0708 C 
 11 11 1 3 Sandstrom, Miss. Marguerite Rut female 4 1 1 PP 9549 16.7000 G6 S 
 12 12 1 1 Bonnell, Miss. Elizabeth female 58 0 0 113783 26.5500 C103 S 
 13 13 0 3 Saundercock, Mr. William Henry male 20 0 0 A/5. 2151 8.0500 S 
 14 14 0 3 Andersson, Mr. Anders Johan male 39 1 5 347082 31.2750 S 
 15 15 0 3 Vestrom, Miss. Hulda Amanda Adolfina female 14 0 0 350406 7.8542 S 
 16 16 1 2 Hewlett, Mrs. (Mary D Kingcome) female 55 0 0 248706 16.0000 S 
 17 17 0 3 Rice, Master. Eugene male 2 4 1 382652 29.1250 Q 
 19 19 0 3 Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele) female 31 1 0 345763 18.0000 S 
 21 21 0 2 Fynney, Mr. Joseph J male 35 0 0 239865 26.0000 S 
 22 22 1 2 Beesley, Mr. Lawrence male 34 0 0 248698 13.0000 D56 S 
 23 23 1 3 McGowan, Miss. Anna "Annie" female 15 0 0 330923 8.0292 Q 
 24 24 1 1 Sloper, Mr. William Thompson male 28 0 0 113788 35.5000 A6 S 
 25 25 0 3 Palsson, Miss. Torborg Danira female 8 3 1 349909 21.0750 S 
 26 26 1 3 Asplund, Mrs. Carl Oscar (Selma Augusta Emilia Johansson) female 38 1 5 347077 31.3875 S 
 28 28 0 1 Fortune, Mr. Charles Alexander male 19 3 2 19950 263.0000 C23 C25 C27 S 
 31 31 0 1 Uruchurtu, Don. Manuel E male 40 0 0 PC 17601 27.7208 C 
 34 34 0 2 Wheadon, Mr. Edward H male 66 0 0 C.A. 24579 10.5000 S 
 35 35 0 1 Meyer, Mr. Edgar Joseph male 28 1 0 PC 17604 82.1708 C 
 36 36 0 1 Holverson, Mr. Alexander Oskar male 42 1 0 113789 52.0000 S 
 38 38 0 3 Cann, Mr. Ernest Charles male 21 0 0 A./5. 2152 8.0500 S 
 39 39 0 3 Vander Planke, Miss. Augusta Maria female 18 2 0 345764 18.0000 S 
 ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ ⋮ 
 857 857 1 1 Wick, Mrs. George Dennick (Mary Hitchcock) female 45 1 1 36928 164.8667 S 
 858 858 1 1 Daly, Mr. Peter Denis male 51 0 0 113055 26.5500 E17 S 
 859 859 1 3 Baclini, Mrs. Solomon (Latifa Qurban) female 24 0 3 2666 19.2583 C 
 861 861 0 3 Hansen, Mr. Claus Peter male 41 2 0 350026 14.1083 S 
 862 862 0 2 Giles, Mr. Frederick Edward male 21 1 0 28134 11.5000 S 
 863 863 1 1 Swift, Mrs. Frederick Joel (Margaret Welles Barron) female 48 0 0 17466 25.9292 D17 S 
 865 865 0 2 Gill, Mr. John William male 24 0 0 233866 13.0000 S 
 866 866 1 2 Bystrom, Mrs. (Karolina) female 42 0 0 236852 13.0000 S 
 867 867 1 2 Duran y More, Miss. Asuncion female 27 1 0 SC/PARIS 2149 13.8583 C 
 868 868 0 1 Roebling, Mr. Washington Augustus II male 31 0 0 PC 17590 50.4958 A24 S 
 870 870 1 3 Johnson, Master. Harold Theodor male 4 1 1 347742 11.1333 S 
 871 871 0 3 Balkic, Mr. Cerin male 26 0 0 349248 7.8958 S 
 872 872 1 1 Beckwith, Mrs. Richard Leonard (Sallie Monypeny) female 47 1 1 11751 52.5542 D35 S 
 873 873 0 1 Carlsson, Mr. Frans Olof male 33 0 0 695 5.0000 B51 B53 B55 S 
 874 874 0 3 Vander Cruyssen, Mr. Victor male 47 0 0 345765 9.0000 S 
 875 875 1 2 Abelson, Mrs. Samuel (Hannah Wizosky) female 28 1 0 P/PP 3381 24.0000 C 
 876 876 1 3 Najib, Miss. Adele Kiamie "Jane" female 15 0 0 2667 7.2250 C 
 877 877 0 3 Gustafsson, Mr. Alfred Ossian male 20 0 0 7534 9.8458 S 
 878 878 0 3 Petroff, Mr. Nedelio male 19 0 0 349212 7.8958 S 
 880 880 1 1 Potter, Mrs. Thomas Jr (Lily Alexenia Wilson) female 56 0 1 11767 83.1

## Spilt data to 70% for train and 30% for test

In [6]:
set.seed(42)
n <- nrow(titanic_train)
id <- sample(1:n, n*0.7) ## 70% Train 30% Test
train_data <- titanic_train[id, ]
test_data <- titanic_train[-id, ]

## Train model
#### Select "Sex" variable in dataset to train model.
#### When we write equation as Survived = f(Sex).

In [7]:
titanic_model <- glm(Survived ~ Sex, data = train_data, family = "binomial")
prob_train <- predict(titanic_model, type = "response") ## probability of Survived
train_data$Prob_Survived <- prob_train
train_data$Pred_Survived <- ifelse(train_data$Prob_Survived >= 0.5, 1, 0)

In [8]:
confus_matrix_train <- table(train_data$Pred_Survived, train_data$Survived, dnn = c("Predicted", "Actual"))
confus_matrix_train

Actual
Predicted 0 1
 0 252 63
 1 45 139

## Test model

In [9]:
prob_test <- predict(titanic_model, newdata = test_data, type = "response")
test_data$Prob_Survived <- prob_test
test_data$Pred_Survived <- ifelse(test_data$Prob_Survived >= 0.5, 1, 0)

In [11]:
confus_matrix_test <- table(test_data$Pred_Survived, test_data$Survived, dnn = c("Predicted", "Actual"))
confus_matrix_test

Actual
Predicted 0 1
 0 108 30
 1 19 58

## Evaluate model
### We use confusion matrix to evaluate model between train dataset and test dataset to calculate Accuracy, Precision and Recall.

In [15]:
cat("Accuracy of train:", (confus_matrix_train[1, 1]+confus_matrix_train[2, 2])/sum(confus_matrix_train))
cat("\nPrecision of train :", confus_matrix_train[2, 2]/(confus_matrix_train[2, 1]+confus_matrix_train[2, 2]))
cat("\nRecall of train:", confus_matrix_train[2, 2]/(confus_matrix_train[1, 2]+confus_matrix_train[2, 2])) 

Accuracy of train: 0.7835671
Precision of train : 0.7554348
Recall of train: 0.6881188

In [16]:
cat("Accuracy of test:", (confus_matrix_test[1, 1]+confus_matrix_test[2, 2])/sum(confus_matrix_test))
cat("\nPrecision of test :", confus_matrix_test[2, 2]/(confus_matrix_test[2, 1]+confus_matrix_test[2, 2]))
cat("\nRecall of test :", confus_matrix_test[2, 2]/(confus_matrix_test[1, 2]+confus_matrix_test[2, 2]))

Accuracy of test: 0.772093
Precision of test : 0.7532468
Recall of test : 0.6590909

## Conclusion
### Found that Accuracy of test is near by train at around 77%. Moreover, Precision and Recall of test is close to train.